### PR TITLE
Add forward iterator to Selection

### PR DESCRIPTION
**Pull request by cattabiani**
Lazy forward const_iterator for Selection that yields values without allocating a vector.

- Add const_iterator nested class with begin()/end() (inline, forward-only)
- Add __iter__ Python binding via py::make_iterator
- Add C++ and Python tests
- flatten() unchanged for backward compatibility

Fix: https://github.com/openbraininstitute/libsonata/issues/55

### CODE REVIEW

File: include/bbp/sonata/selection.h [hunk 1]
```diff
@@ -51,6 +52,77 @@ class SONATA_API Selection
...
+            : range_it_(range_it)
+            , range_end_(range_end)
+            , current_(range_it != range_end ? std::get<0>(*range_it) : 0) { }
+
+        Ranges::const_iterator range_it_{};
+        Ranges::const_iterator range_end_{};
+        Value current_ = 0;
+    };
+
+    /**
```
Comment (mgeplf) [inline include/bbp/sonata/selection.h:119]: no need for these comments, I would hope ppl know what begin/end are

File: python/tests/test_population.py [hunk 2]
```diff
@@ -104,6 +104,35 @@ def test_union_intersection(self):
...
+        self.assertEqual(list(sel), sel.flatten().tolist())
+
+        # Empty selection yields nothing
+        empty = Selection([])
+        self.assertEqual(list(empty), [])
+
+        # Order preserved (not sorted)
+        sel2 = Selection([(10, 12), (5, 8), (0, 2)])
+        self.assertEqual(list(sel2), [10, 11, 5, 6, 7, 0, 1])
+        self.assertEqual(list(sel2), sel2.flatten().tolist())
```
Comment (mgeplf) [inline python/tests/test_population.py]: this doesn't check anything that is from the implentation, imo, nor does the rest; we have to assume that pybind does the proper iterator binding

File: python/bindings.cpp [hunk 3]
```diff
@@ -492,18 +492,23 @@ PYBIND11_MODULE(_libsonata, m) {
...
-                               fmt::join(ranges.end() - 3, ranges.end(), ", "));
-        });
+                 return fmt::format("Selection([{}, ..., {}])",
+                                    fmt::join(ranges.begin(), ranges.begin() + 3, ", "),
+                                    fmt::join(ranges.end() - 3, ranges.end(), ", "));
+             })
+        .def(
+            "__iter__",
+            [](const Selection& obj) { return py::make_iterator(obj.begin(), obj.end()); },
+            py::keep_alive<0, 1>());
```
Comment (mgeplf) [inline python/bindings.cpp:511]: why do you need the `py::keep_alive`?

File: tests/test_selection.cpp [hunk 4]
```diff
@@ -31,6 +31,65 @@ TEST_CASE("Selection", "[base]") {
...
+            std::vector<uint64_t> collected(sel.begin(), sel.end());
+            CHECK(collected.empty());
+        }
+        // Single range
+        {
+            const auto sel = Selection({{3, 7}});
+            std::vector<uint64_t> collected(sel.begin(), sel.end());
+            CHECK(collected == std::vector<uint64_t>{3, 4, 5, 6});
+        }
+        // Multiple ranges — order preserved (not sorted)
```
Comment (mgeplf) [inline tests/test_selection.cpp:48]: duplicate of `Matches flatten() output`

File: tests/test_selection.cpp [hunk 5]
```diff
@@ -31,6 +31,65 @@ TEST_CASE("Selection", "[base]") {
...
+            CHECK(collected == std::vector<uint64_t>{3, 4, 0, 1, 2});
+        }
+        // Matches flatten() output
+        {
+            const auto sel = Selection({{10, 12}, {5, 8}, {0, 2}});
+            std::vector<uint64_t> collected(sel.begin(), sel.end());
+            CHECK(collected == sel.flatten());
+        }
+        // Range-based for loop
+        {
```
Comment (mgeplf) [inline tests/test_selection.cpp]: how would the above work but this fail?

File: tests/test_selection.cpp [hunk 6]
```diff
@@ -31,6 +31,65 @@ TEST_CASE("Selection", "[base]") {
...
+            CHECK(collected == std::vector<uint64_t>{1, 2, 3});
+        }
+        // Post-increment returns previous value
+        {
+            const auto sel = Selection({{5, 8}});
+            auto it = sel.begin();
+            CHECK(*it++ == 5);
+            CHECK(*it == 6);
+        }
+        // Iterator matches flatten() after union (sorted/merged ranges)
```
Comment (mgeplf) [inline tests/test_selection.cpp]: what is `Iterator matches flatten() after union (sorted/merged ranges)` and `Iterator matches flatten() after intersection`? they don't depend on the iterator implementation, do they?

File: include/bbp/sonata/selection.h [hunk 7]
```diff
@@ -51,6 +52,77 @@ class SONATA_API Selection
...
+        }
+
+        const_iterator operator++(int) {
+            auto tmp = *this;
+            ++(*this);
+            return tmp;
+        }
+
+        bool operator==(const const_iterator& other) const {
+            return range_it_ == other.range_it_;
```
Comment (mgeplf) [inline include/bbp/sonata/selection.h]: seems strange, why doesn't this depend on `current_`? two iterators could compare equal just if they happen to point to the same `range_it_`?

File: include/bbp/sonata/selection.h (hunk 1 above)
Comment (cattabiani) [reply include/bbp/sonata/selection.h:119]: removed

File: python/tests/test_population.py (hunk 2 above)
Comment (cattabiani) [reply python/tests/test_population.py]: The C++ iterator implementation is tested in the C++ test suite. The Python test here is a thin integration check that the pybind11 `__iter__` binding is wired up correctly (i.e., `py::make_iterator(begin(), end())` is present and functional). There's no Python-side logic to test beyond confirming pybind11 does its job. A single `assertEqual(list(sel), sel.flatten().tolist())` would suffice as a smoke test. Ultimately the rest is redundant but harmless

File: python/bindings.cpp (hunk 3 above)
Comment (cattabiani) [reply python/bindings.cpp:511]: The keep_alive<0, 1> ensures the Selection object stays alive as long as the iterator exists. Without it:

```
it = iter(Selection([(0, 100)]))  # Selection is a temporary, refcount → 0
gc.collect()                       # C++ Selection destroyed
next(it)                           # iterator dereferences freed memory → UB
```

py::make_iterator(begin(), end()) captures raw C++ iterators that point into the Selection's internal storage. If Python GCs the Selection while the iterator is still alive, those iterators dangle. keep_alive<0, 1> ties the Selection's lifetime to the iterator's, preventing that.

It's standard pybind11 practice for any make_iterator call where the iterator doesn't own its data. I have a test (test_keep_alive.py) that demonstrates the crash under ASan if you remove it:

```
"""
Test that demonstrates the need for keep_alive<0, 1> on Selection.__iter__.

Without keep_alive, the Selection object can be garbage-collected while
the iterator still holds dangling C++ iterators into its internal data.
This test tries to trigger that scenario.
"""

import gc
import libsonata


def test_dangling_iterator():
    """Iterator from a temporary Selection should not crash."""
    # Get an iterator from a Selection that immediately goes out of scope
    it = iter(libsonata.Selection([(0, 100)]))

    # Drop all references and force GC
    gc.collect()
    gc.collect()
    gc.collect()

    # Allocate some memory to increase chance the freed region is overwritten
    junk = [list(range(1000)) for _ in range(100)]

    # Now iterate — if Selection was collected, these are dangling pointers
    values = list(it)
    expected = list(range(100))
    assert values == expected, f"Got {values[:10]}... expected {expected[:10]}..."


def test_dangling_iterator_next():
    """Calling next() one-by-one after GC."""
    it = iter(libsonata.Selection([(50, 60)]))
    gc.collect()
    gc.collect()
    gc.collect()

    # Overwrite freed memory
    junk = [bytearray(4096) for _ in range(50)]

    results = []
    for _ in range(10):
        results.append(next(it))

    assert results == list(range(50, 60)), f"Got {results}"


if __name__ == "__main__":
    test_dangling_iterator()
    print("test_dangling_iterator PASSED")
    test_dangling_iterator_next()
    print("test_dangling_iterator_next PASSED")
```

fails with:

```
katta@coeus libsonata % python python/tests/test_keep_alive.py

Traceback (most recent call last):
  File "/Users/katta/OBI/libsonata/python/tests/test_keep_alive.py", line 50, in <module>
    test_dangling_iterator()
  File "/Users/katta/OBI/libsonata/python/tests/test_keep_alive.py", line 29, in test_dangling_iterator
    assert values == expected, f"Got {values[:10]}... expected {expected[:10]}..."
           ^^^^^^^^^^^^^^^^^^
AssertionError: Got [0]... expected [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]...
```

File: tests/test_selection.cpp (hunk 4 above)
Comment (cattabiani) [reply tests/test_selection.cpp:48]: match flatten is a cross-check. This is a check vs reference. If both methods change or have a bug we would not notice with match flatten alone. Practically, they cover almost the same stuff and the probability of accidentally changing both is very low

Comment (cattabiani) [reply tests/test_selection.cpp:48]: ... ergo, I can remove it if you insist

File: tests/test_selection.cpp (hunk 5 above)
Comment (cattabiani) [reply tests/test_selection.cpp]: yes, I went a little too far with the tests. Removed

File: tests/test_selection.cpp (hunk 6 above)
Comment (cattabiani) [reply tests/test_selection.cpp]: it is true that these test iterators as well as union and intersection. Maybe I can move them. These are probably superfluous and mostly a sanity check. Since union and intersection create new ranges and these ranges are ordered I wanted to make sure that the order was preserved among methods

I can also remove them, as you prefer

File: include/bbp/sonata/selection.h (hunk 7 above)
Comment (cattabiani) [reply include/bbp/sonata/selection.h]: you are partially right. We need to compare both range and current:

- compare just current: selection could have overlapping range where values are equal
- compare just range: we could be on the same range, different value position

fortunately, over a single range values are unique. We can compare them

Small caveat: when we reach the end, for efficiency, we do not reset current. This means that a simple double comparison (range + current) would result false for end pointers that should compare equal

I added a couple of tests and fixed

File: python/tests/test_population.py (hunk 2 above)
Comment (mgeplf) [reply python/tests/test_population.py]: keeping `self.assertEqual(list(sel2), [10, 11, 5, 6, 7, 0, 1])` makes sense, but the rest are just redundant, imo, and don't show anything beyond that `__iter__` works.  Best to remove them, imo.

File: python/bindings.cpp (hunk 3 above)
Comment (mgeplf) [reply python/bindings.cpp:511]: Weird, I get:
```
python3 it.py
test_dangling_iterator PASSED
test_dangling_iterator_next PASSED
```
Which is what I had tried before, I didn't do it with ASAN though, let me see if that does it.  How are you compiling it with that?

File: tests/test_selection.cpp (hunk 4 above)
Comment (mgeplf) [reply tests/test_selection.cpp:48]: ok; keep it for now

File: python/tests/test_population.py (hunk 2 above)
Comment (cattabiani) [reply python/tests/test_population.py]: ok

File: tests/test_selection.cpp (hunk 6 above)
Comment (mgeplf) [reply tests/test_selection.cpp]: I may have to reset my expectations, but normally I think of tests as trying to cover specific behavior of a particular feature to test its limitations, or edge cases.  By having tests that exercise things that aren't directly related, it signals to me that I need to look at how the union/intersection uses the iterators (event hough they don't.)

Just my 2 cents, if you think these are helpful, I guess I'll recalibrate.

File: python/bindings.cpp (hunk 3 above)
Comment (cattabiani) [reply python/bindings.cpp:511]:
```
# Clean build with sanitizers
cmake -B build-asan \
  -DCMAKE_CXX_FLAGS="-fsanitize=address -fno-omit-frame-pointer" \
  -DCMAKE_C_FLAGS="-fsanitize=address -fno-omit-frame-pointer" \
  -DCMAKE_EXE_LINKER_FLAGS="-fsanitize=address" \
  -DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=address" \
  -DSONATA_PYTHON=ON \
  .

cmake --build build-asan -j$(nproc)
```

```
PYTHONPATH=build-asan/python python3 python/tests/test_keep_alive.py
```

you should see something like:

```
==12345==ERROR: AddressSanitizer: heap-use-after-free on address 0x...
```

the problem is that the failure is non-deterministic in general. It us UB

File: tests/test_selection.cpp (hunk 6 above)
Comment (cattabiani) [reply tests/test_selection.cpp]: ok, removed. It is true it is more of an integration test

File: python/bindings.cpp (hunk 3 above)
Comment (mgeplf) [reply python/bindings.cpp:511]: Yup, nice.  Thanks.

**Review by mgeplf — APPROVED** [review]
